Sprint 8 end task.

IvanSimonovDev opened the pull request:
Realize:
1) time and duration of tasks
2) providing prioritized tasks list
3) time-checking validation
4) using Stream API in code
5) many tests

IvanSimonovDev committed:
Changes:
1) Add temporal fields and method getEndTime() to Task and it's subclasses.
2) Add additional constructors with temporal fields to Task and Subtask.
3) Change method clone() in Task and it's subclasses.
4) Add automatic calculation of Epic's duration and start time.
5) Change serialization/deserialization due to new functionality.
6) Cover new functionality with tests.

IvanSimonovDev committed:
Changes:
1) Add comparator for all tasks.
2) Realize prioritized tasks ordered set.
3) Realize validation of tasks' start time and duration.
4) Replace foreach usings in managers with streams.

IvanSimonovDev committed:
Changes:
1) Replace ManagerSaveException with ManagerSaveLoadException.
2) Changed exception processing.
3) Changed updateEpic() method in InMemoryTaskManager.
4) Update tests.

VladimirTitov4 commented on src/lib/FileBackedTaskManager.java:
```diff
@@ -134,20 +134,25 @@ private void saveEveryTask(Writer writer) throws java.io.IOException {
+                            task.title, task.description, task.status.toString(), "-",
+                            stringStartTime, stringDuration, "\n");
                     break;
                 case "lib.tasks.Epic":
```
Желательно работать с enum TaskType

VladimirTitov4 commented on src/lib/InMemoryTaskManager.java:
```diff
@@ -64,13 +80,16 @@ public ArrayList<SubTask> getSubTasksList() {
     @Override
     public void deleteAllSubTasks() {
-        for (SubTask subTask : subTaskStorage.values()) {
+        Consumer<SubTask> cleanFunction = subTask -> {
```
Необычная реализация, а чем это лучше обычного stream().forEach(  ?

VladimirTitov4 commented on src/lib/InMemoryTaskManager.java:
```diff
@@ -171,16 +200,42 @@ protected void setEpicStatus(short epicId) {
+        List<SubTask> subTasksOfEpic = this.subTasksOfEpic(epic.id);
+
+        epic.duration = Duration.ofMinutes(0);
+        Predicate<SubTask> filterTemporalNulls = element -> (element.duration != null) && (element.startTime != null);
```
Интересно реализовали с предикатом

IvanSimonovDev committed:
Changes:
1) Replace in streams .peek(...).toList() with forEach();
2) FileBackedTaskManager#saveEveryTask(Writer writer):
   Replace switch based on canonical name of class with switch based on enumeration TaskType;
